### ISSUE TITLE
Android installation problem ? 

### ISSUE THREAD

**Issue by RoyiNamir** [mobilemindtec/nativescript-background-task#9]
I might be missing something : 

If  I try to install it via npm  :  

![image](https://user-images.githubusercontent.com/2842123/33715786-705acce2-db5c-11e7-96c3-dbe014d221aa.png)

Also : 
I've followed all your instructions : 

[is the include.gradle](https://gist.github.com/4a601d0bd8a8e637f0db673eb8b0fac2) 

[this is the app.gradle](https://gist.github.com/anonymous/c5dfca0a568052035c7be6c174f0b88d)

Also remove and added platforms.

this is the structure : 

![image](https://user-images.githubusercontent.com/2842123/33715901-dbaf8500-db5c-11e7-963c-e0365fc1d7cf.png)

I've imported in code : 

    var BackgroundTask = require("nativescript-background-task")

but when I run `tns run android` : 

![image](https://user-images.githubusercontent.com/2842123/33715966-17c323f8-db5d-11e7-8612-f689fed828c4.png)

In node_nodules :  I don't have the package : 

![image](https://user-images.githubusercontent.com/2842123/33715994-2b1e891a-db5d-11e7-9755-6f40eaa5505d.png)

**Question**

what am I missing ?

**Comment by mobilemindtec** [comment]
Ho @RoyiNamir 

This plugin is not registered on npm. To use, you can add github repo directly on your package.json and so run npm install.

```
dependencies: {
    "nativescript-background-task": "https://github.com/mobilemindtec/nativescript-background-task.git"
}
```

**Comment by RoyiNamir** [comment]
@mobilemindtec  
❤❤❤ Thank you !Thank you !Thank you !Thank you !  
One last question please : 

Any progress callback ( for progress bar) ?

**Comment by mobilemindtec** [comment]
@RoyiNamir 

No, don't has progress callback, has only finished callback. You can use `ActivityIndicator` to show indeterminate progress, I usually use it.

**Comment by RoyiNamir** [comment]
Ok . anyway - I've tried download a demo file : 
```
  BackgroundTask.getFile({
            url: 'http://www.sample-videos.com/doc/Sample-doc-file-100kb.doc',
            toFile: "/raw/new.jpg",
            identifier: 1,
            partBytesSize: 0, // use default 2MB
            checkPartialDownload: true,
            headers: [
            
            ],
            doneCallback: function(identifier){
               console.log(identifier)
            },
            errorCallback: function(error){
                console.log(error);
            }
        })

```

but I get an error : 

![image](https://user-images.githubusercontent.com/2842123/33716916-99346908-db60-11e7-8719-9e69fc89967b.png)

**Comment by mobilemindtec** [comment]
Can you make the sources files available for this demo project?

**Comment by RoyiNamir** [comment]
Sure .

https://ufile.io/sbep5


Just use android platform and run.

then youll see this : 

![image](https://user-images.githubusercontent.com/2842123/33717309-18716dbe-db62-11e7-92e7-a42605b4dcab.png)


just click on the download file ( it downloads a dummy file from web)

Thank you very much for your effort.

**Comment by mobilemindtec** [comment]
@RoyiNamir,

Some points:

You shoud set a valid path to `toFile` property.  Android lib is throwing `java.io.FileNotFoundException: /raw/new.jpg.part: open failed: ENOENT (No such file or directory)`. File don't need exists, but path need exists.

In error case, you can check with:

```
errorCallback: function(error){
    console.log("Download error: " + error[0]);  // identifier
   console.log("Download error: " + error[1]); // error message
}
```
Changes in your code to work:
```
        var tmp = knownFolders.temp() // add here

        BackgroundTask.getFile({
            url: 'http://www.sample-videos.com/doc/Sample-doc-file-100kb.doc',
            toFile: tmp.path + "/Sample-doc-file-100kb.doc", //  chenge here
            identifier: 1,
            partBytesSize: 0, // use default 2MB
            checkPartialDownload: true,
            headers: [
            
            ],
            doneCallback: function(identifier){
               console.log("Donwload sucesss: " + identifier)
            },
            errorCallback: function(error){
                console.log("Download error: " + error[1]);
            }
        })
```

But I do not can get your related error. `cannot read property 'task' of undefined` indicates compile problem, as if the lib android had not been compiled correctly with the nativescript project.

Steps to test:

* Download project
* run `npm install`
* run `tns platform add android@3.3.1`
* run `tns run android`

Result:

![screenshot from 2017-12-07 12-10-54](https://user-images.githubusercontent.com/8561714/33719402-b8c5b5ec-db47-11e7-8371-fa37442da4c4.png)

Maybe you should clean your project `cd platforms/android && ./gradlew clean` and run again to resolve this error.

**Comment by RoyiNamir** [comment]
It does work now

But why do you say that `raw` folder is not exists ? 

I do have it here : 

![image](https://user-images.githubusercontent.com/2842123/33719893-e5a27fd2-db6a-11e7-8f10-1fcbc3fe4a1e.png)


If I play a song from the app it does read the `raw/a2.mp3` file ( in the `raw`) folder

**Comment by mobilemindtec** [comment]
`/raw/new.jpg`is absolute path and `raw/new.jpg` is relative path.

**Comment by RoyiNamir** [comment]
I know that - but AFAIK `app ` is the root -  so `/raw` should be fine . no ? ( I know that it's not working with `/`) - I'm just trying to see what' am I missing.

In the POV of android  - what is the root ? and does `/raw` in my project is copied there to the root also 
?

**Comment by mobilemindtec** [comment]
No, root of app não is android root. I believe that apps do not have access to the root of the android, for security reasons. Root of your app is `/data/data/org.nativescript.Groceries/files/app`

**Comment by RoyiNamir** [comment]
Ok. please notice that your plugin doesnt work with @next versionof nativescript
it will be wonderful to allow that
And thank you for your knowledge and kindness

**Comment by mobilemindtec** [comment]
@RoyiNamir,

@next version is not official release. Official release android version is 3.3.1. I believe that @next has any bug:

When I try add @next on project:

![screenshot from 2017-12-07 14-19-15](https://user-images.githubusercontent.com/8561714/33725558-a45845ea-db59-11e7-82ff-64cdeb0ea342.png)

It's always good to help!

**Comment by RoyiNamir** [comment]
Please - last question  : 
What's the difference between : 
```
  console.log(knownFolders.currentApp().path)
  console.log(knownFolders.documents().path)
```


the result is : 
```
JS: /data/data/org.nativescript.Groceries/files/app
JS: /data/user/0/org.nativescript.Groceries/files
```
when should I use which ? 

for example : when you unzip - where should it be saved ? I can't understand the difference :

**Comment by mobilemindtec** [comment]
@RoyiNamir,

`currentApp().path`: Gets the root folder for the current application. This Folder is private for the application and not accessible from Users/External apps. iOS - this folder is read-only and contains the app and all its resources.

`documents()`: Gets the Documents folder available for the current application. This Folder is private for the application and not accessible from Users/External apps.

https://docs.nativescript.org/api-reference/modules/_file_system_.knownfolders

Basically, on `knownFolders.currentApp()` folder you can only read files like video, images, etc.. and `knownFolders.documents()` folder you can create sub folders to save files. 

In zip case, the better is you create a temp folder using ` knownFolders.temp().path` to do work, and after done, save files on `knownFolders.documents` or on sdcard. That's how I do it.

**Comment by RoyiNamir** [comment]
But From my testings - I **can(!!!)** write to  `fs.knownFolders.currentApp() ` :  - my phone is not rooted

```
var documents = fs.knownFolders.currentApp() ;
        var path = fs.path.join(documents.path, "/raw/777.txt");
        var file = fs.File.fromPath(path);
      file.writeText("Something")
```
File `777.txt` is created

Look what happens when I scan both `currentApp()` and `documents()`
```
JS: Files from.currentApp().getFolder("raw")
JS: 777.txt

JS: Files from .documents().getFolder("app/raw")
JS: 777.txt
```

So why did you say : 

"_Basically, on knownFolders.currentApp() folder you can only read files like video_"

?thank you

**Comment by mobilemindtec** [comment]
Yes, I know. But you want make your app multiplatform you should not do this, your code will break on IOS. 

currentApp documentation say:

_iOS - this folder is read-only and contains the app and all its resources._

**Comment by RoyiNamir** [comment]
What can I say - you're the man.
I have no words to thank you for this day's help.

Thank you x1000 times.
bye

**Comment by mobilemindtec** [comment]
thank man!
